### PR TITLE
Add ignore-pattern option to no-unused-variable

### LINKED ISSUE #314: Selectively suppres unused variable warning
Haskell has a nice way to selectively suppress their equivalent of *unused variable* by prepending the variable with an underscore. I'm particularly interested in suppressing the warning for unused function arguments.

### PR DESCRIPTION
**Pull request by Pajn**
fixes #314

### CODE REVIEW

**Comment by JKillian** [comment]
Thanks @Pajn! Will review in the next couple of days.

As a sidenote, we might want to think about a similar option for #522 as well.

File: README.md [hunk 1]
```diff
@@ -200,6 +200,7 @@ A sample configuration file with all options is available [here](https://github.
     * `"check-parameters"` disallows unused function and constructor parameters.
         * NOTE: this option is experimental and does not work with classes that use abstract method declarations, among other things. Use at your own risk.
     * `"react"` relaxes the rule for a namespace import named `React` (from either the module `"react"` or `"react/addons"`). Any JSX expression in the file will be treated as a usage of `React` (because it expands to `React.createElement`).
+    * `{"ignore-pattern": "pattern"}` where pattern is a case sensitive regexp. Variable names that matches the pattern will be ignored.
```
Comment (JKillian) [inline README.md]: I wish we could accept a literal regex, but since we're using JSON, I suppose we have no good way to do this.

Comment (JKillian) [inline README.md]: Just a couple of grammar tweaks would be good here:

* "case sensitive" => "case-sensitive"
* "matches" => "match"

File: src/rules/noUnusedVariableRule.ts [hunk 2]
```diff
@@ -279,8 +282,18 @@ class NoUnusedVariablesWalker extends Lint.RuleWalker {
...
-        if (highlights == null || highlights[0].highlightSpans.length <= 1) {
+        if ((highlights == null || highlights[0].highlightSpans.length <= 1) && !this.isIgnored(name)) {
             this.addFailure(this.createFailure(position, name.length, `${Rule.FAILURE_STRING}'${name}'`));
         }
     }
+
+    private isIgnored(name: string) {
+        const options = this.getOptions();
+        for (const option of options) {
+          if (typeof option === "object") {
```
Comment (JKillian) [inline src/rules/noUnusedVariableRule.ts]: Indentation is off for this `if` block

File: src/rules/noUnusedVariableRule.ts [hunk 3]
```diff
@@ -279,8 +282,18 @@ class NoUnusedVariablesWalker extends Lint.RuleWalker {
...
+            const {"ignore-pattern": ignorePattern} = option;
+            return ignorePattern && new RegExp(ignorePattern).test(name);
```
Comment (JKillian) [inline src/rules/noUnusedVariableRule.ts]: We normally use an explicit check here, i.e. `ignorePattern != null`

**Comment by JKillian** [comment]
Looks solid to me @Pajn, left a few notes for grammar/style tweaks

**Comment by Pajn** [comment]
Fixed

**Comment by JKillian** [comment]
Thanks! Merged

File: src/rules/noUnusedVariableRule.ts [hunk 4]
```diff
@@ -279,8 +282,20 @@ class NoUnusedVariablesWalker extends Lint.RuleWalker {
...
         }
     }
+
+    private isIgnored(name: string) {
+        const options = this.getOptions();
+        for (const option of options) {
+            if (typeof option === "object") {
+                const {"ignore-pattern": ignorePattern} = option;
+                if (ignorePattern != null) {
+                    return new RegExp(ignorePattern).test(name);
```
Comment (wereHamster) [inline src/rules/noUnusedVariableRule.ts]: What performance overhead does compiling the pattern on each check have? I'd expect this function to be called very frequently...

Comment (Pajn) [inline src/rules/noUnusedVariableRule.ts]: It will be falled once per unused variable, to be noticeable you would need loads of unused variables.

Comment (JKillian) [inline src/rules/noUnusedVariableRule.ts]: @wereHamster This is a fair point which I considered and we can change it if needed so the RegExp is just compiled once and stored as a class member. However, my hunch is that compared to `languageService.getDocumentHighlights` the impact is small. Have you been running into performance issues in general?

Comment (adidahiya) [inline src/rules/noUnusedVariableRule.ts]: regardless, recompiling the regexp is unnecessary; let's move it to a class member.

**Pull request merged**